### PR TITLE
NVMe-over-fabrics support, part 3

### PR DESCRIPTION
**Pull request by no92**

### CODE REVIEW

File: kernel/eir/protos/uefi/entry.cpp [hunk 1]
```diff
@@ -90,6 +118,128 @@ initgraph::Task findDtb{
...
+
+	    auto status = bs->handle_protocol(loadedImage->device_handle, &pxe_guid, reinterpret_cast<void **>(&pxe));
+		if(status != EFI_SUCCESS)
+			return;
+
+		EFI_CHECK(bs->allocate_pool(EfiLoaderData, sizeof(pxe_info), reinterpret_cast<void **>(&pxeInfo)));
+		new (pxeInfo) pxe_info{};
+
+		// TODO: support IPv6
+		assert(pxe->mode->using_ipv6 == false);
```
Comment (avdgrinten) [inline kernel/eir/protos/uefi/entry.cpp:195]: We can probably fail gracefully here by just returning from the lambda (or similar).

File: kernel/eir/protos/uefi/entry.cpp [hunk 2]
```diff
@@ -90,6 +118,128 @@ initgraph::Task findDtb{
...
+
+		eir::infoLogger() << "eir: PXE available, " << (pxe->mode->started ? "started" : "stopped") << frg::endlog;
+
+		assert(pxe->mode->started);
```
Comment (avdgrinten) [inline kernel/eir/protos/uefi/entry.cpp:199]: Likewise.

File: kernel/eir/protos/uefi/entry.cpp [hunk 3]
```diff
@@ -90,6 +118,128 @@ initgraph::Task findDtb{
...
+			if(!pxeInfo->gateway_ip.addr[0]) {
+				pxeInfo->gateway_ip = {};
+				memcpy(&pxeInfo->gateway_ip.v4, pxe->mode->dhcp_ack.dhcpv4.bootp_gi_addr, sizeof(pxeInfo->gateway_ip.v4));
+			}
+		} else {
+			pxeInfo->gateway_ip.addr[0] = convertIp(gateway_str);
+		}
+
+		// TODO: fall back to using DHCP option 54 or DHCP next-server
+		assert(pxeInfo->server_ip.addr[0]);
```
Comment (avdgrinten) [inline kernel/eir/protos/uefi/entry.cpp:254]: Likewise.

File: kernel/eir/protos/uefi/entry.cpp [hunk 4]
```diff
@@ -148,6 +153,44 @@ initgraph::Task preparePxe{
...
 		new (pxeInfo) pxe_info{};
 
+		EFI_CHECK(bs->handle_protocol(loadedImage->device_handle, &devpath_guid, reinterpret_cast<void **>(&devpath)));
+		EFI_CHECK(bs->locate_protocol(&devpath2text_guid, nullptr, reinterpret_cast<void **>(&devpath2text)));
+
+		auto devpathstr = devpath2text->convert_device_path_to_text(devpath, true, true);
+		assert(devpathstr);
+
+		char *devpathascii = nullptr;
+		{
```
Comment (avdgrinten) [inline kernel/eir/protos/uefi/entry.cpp:163]: Braces here are unnecessary.

File: kernel/thor/system/pci/pci_acpi.cpp [hunk 5]
```diff
@@ -245,10 +245,11 @@ static initgraph::Task discoverAcpiRootBuses{&globalInitEngine, "pci.discover-ac
 		uacpi_find_devices_at(
 			uacpi_namespace_get_predefined(UACPI_PREDEFINED_NAMESPACE_SB),
 			pciRootIds, [](void*, uacpi_namespace_node *node, uint32_t) {
-				uint64_t seg = 0, bus = 0;
+				uint64_t seg = 0, bus = 0, uid = 0;
 
 				uacpi_eval_simple_integer(node, "_SEG", &seg);
 				uacpi_eval_simple_integer(node, "_BBN", &bus);
+				uacpi_eval_simple_integer(node, "_UID", &uid);
```
[the remark this answers is not in the document]
Comment (no92) [reply kernel/thor/system/pci/pci_acpi.cpp:252]: If the call fails, the out value is untouched and a error status is returned.

File: kernel/thor/system/acpi/acpi.cpp [hunk 6]
```diff
@@ -10,10 +10,9 @@
 
 namespace thor::acpi {
 
-coroutine<void> AcpiObject::run() {
+coroutine<void> AcpiObject::run(Properties acpi_properties) {
```
[the remark this answers is not in the document]
Comment (no92) [reply kernel/thor/system/acpi/acpi.cpp:13]: The UID field might not necessarily be present for all devices, and this design allows for future ACPI devices that we choose to expose over mbus to define more properties than the `AcpiObject` class itself sets.

[the remark this answers is not in the document]
Comment (avdgrinten) [reply kernel/thor/system/acpi/acpi.cpp:13]: I'd probably prefer some other way to override this, but that can be added when we need it.

**Review by avdgrinten — APPROVED** [review]

**Pull request merged**